lib/mongoid/clients.rb

```diff
@@ -61,7 +61,12 @@ def disconnect
       #
       # @since 3.0.0
       def with_name(name)
-        clients[name.to_sym] ||= Clients::Factory.create(name)
+        name_as_symbol = name.to_sym
+        return clients[name_as_symbol] if clients[name_as_symbol]
+        CREATE_LOCK.synchronize do
+          return clients[name_as_symbol] if clients[name_as_symbol]
+          clients[name_as_symbol] = Clients::Factory.create(name)
```

Comment:
But any thread (1) waiting to access a certain client would have to wait for the lock to be released if another thread (2) were creating a client (even if the thread (2) were creating a client different from the one the thread (1) were waiting for).